### ISSUE TITLE
[TF 2.0] Model not converging when trained with custom training loop

### ISSUE THREAD

**Issue by FelixWiewel** [tensorflow/tensorflow#28901]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.02 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Binary from Pypi
- TensorFlow version (use command below): 2.0 alpha0
- Python version: 3.6.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0
- GPU model and memory: GTX1070 8GB

**Describe the current behavior**

I have created a simple regression model and trained it using keras and also using a custom training loop. What I noticed is that the model only converges when using the keras fit function. If I use the custom training loop, which manually computes gradients and applies them via an optimizer, the model does not converge. The same problem occurred not only on this very simple regression task but also on a model trained on the UTK Faces data set for age regression.

**Describe the expected behavior**

I would expect both, the model trained with keras fit and the model trained with a custom training loop, to converge.

**Code to reproduce the issue**

```
import tensorflow as tf
import matplotlib.pyplot as plt
import numpy as np

# Define constants
N_train_samples = 600
N_validation_samples = 100
N_evaluation_samples = 100
noise_sig = 0.1
N_epochs = 100
random_seed = 100
batch_size = 16
learning_rate = 0.1

# Create target signal
x = np.linspace(0.0, 3.0, N_train_samples+N_validation_samples+N_evaluation_samples, dtype=np.float32)
y = np.sin(1 + x*x) + noise_sig*np.random.randn(N_train_samples+N_validation_samples+N_evaluation_samples).astype(np.float32)
y_true = np.sin(1 + x*x)


# Define model
class MyModel(object):
    def __init__(self):
        # Create model variables
        self.W0 = tf.Variable(tf.random.normal([1, 10]), name="W0")
        self.b0 = tf.Variable(tf.zeros(10), name="b0")
        self.W1 = tf.Variable(tf.random.normal([10, 1]), name="W1")
        self.b1 = tf.Variable(tf.zeros(1), name="b1")
        self.trainable_variables = [self.W0, self.b0, self.W1, self.b1]

    def __call__(self, inputs):
        output = tf.reshape(inputs, [-1, 1])
        output = tf.nn.sigmoid(tf.add(tf.matmul(output, self.W0), self.b0))
        return tf.add(tf.matmul(output, self.W1), self.b1)


class MyKerasModel(tf.keras.Model):
    def __init__(self):
        super(MyKerasModel, self).__init__()
        # Create model variables
        self.dense0 = tf.keras.layers.Dense(10, activation="sigmoid")
        self.dense1 = tf.keras.layers.Dense(1, activation="linear")

    def call(self, inputs):
        output = tf.reshape(inputs, [-1, 1])
        output = self.dense0(output)
        return self.dense1(output)



# Define training step
@tf.function
def train_step(model, optimizer, x, y):
    with tf.GradientTape() as tape:
        tape.watch(model.trainable_variables)
        y_pred = model(x)
        loss_val = tf.reduce_mean(tf.square(y-y_pred))
    grads = tape.gradient(loss_val, model.trainable_variables)
    optimizer.apply_gradients(zip(grads, model.trainable_variables))
    return loss_val


# Shuffle and partition data set
shuffle_idx = np.arange(0, N_train_samples+N_validation_samples+N_evaluation_samples)
np.random.shuffle(shuffle_idx)
x_shuffled = x[shuffle_idx]
y_shuffled = y[shuffle_idx]
x_train = x_shuffled[0:N_train_samples]
y_train = y_shuffled[0:N_train_samples]
x_validation = x_shuffled[N_train_samples:N_train_samples+N_validation_samples]
y_validation = y_shuffled[N_train_samples:N_train_samples+N_validation_samples]
x_evaluation = x_shuffled[N_train_samples+N_validation_samples:N_train_samples+N_validation_samples+N_evaluation_samples]
y_evaluation = y_shuffled[N_train_samples+N_validation_samples:N_train_samples+N_validation_samples+N_evaluation_samples]

# Create data sets
train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train)).shuffle(N_train_samples).batch(batch_size).repeat()
validation_ds = tf.data.Dataset.from_tensor_slices((x_validation, y_validation)).batch(N_validation_samples)
evaluation_ds = tf.data.Dataset.from_tensor_slices((x_evaluation, y_evaluation)).batch(N_evaluation_samples)

# Create model and optimizer
mdl = MyModel()
opt = tf.optimizers.Adam(learning_rate)

keras_mdl = MyKerasModel()
keras_mdl.compile(optimizer=tf.optimizers.Adam(learning_rate), loss="mse", metrics=["mae"])

# Train model
epoch = 0
train_iters = 0
train_loss = 0.0
for x_feed, y_feed in train_ds:
    train_loss += train_step(mdl, opt, x_feed, y_feed)
    train_iters += 1
    if (train_iters >= int(N_train_samples/batch_size)):
        for x_feed, y_feed in validation_ds:
            y_pred = mdl(x_feed)
            validation_loss = tf.reduce_mean(tf.square(y-y_pred))
        print("Epoch: {} Train loss: {:.5} Validation loss: {:.5}".format(epoch, train_loss/train_iters, validation_loss))
        train_iters = 0
        train_loss = 0.0
        epoch += 1
    if (epoch == N_epochs):
        break

keras_mdl.fit_generator(train_ds, epochs=N_epochs, steps_per_epoch=int(N_train_samples/batch_size),\
        validation_data=validation_ds, validation_steps=1)

# Predict with model and plot results
y_pred = mdl(x)
y_keras_pred = keras_mdl(x)
plt.plot(x, y)
plt.plot(x, y_true)
plt.plot(x, y_pred.numpy())

```

**Comment by muddham** [comment]
@FelixWiewel Able to reproduce the issue with the provided code and visualized the graph

**Comment by llan-ml** [comment]
Hi @FelixWiewel What if using `MyKerasModel` but with the custom training loop?

**Comment by AlexanderBartler** [comment]
@FelixWiewel @llan-ml 
I can reproduce the bug as well and already tried to use the `MyKerasModel` with the same wrong behavior.

**Comment by FelixWiewel** [comment]
> Hi @FelixWiewel What if using `MyKerasModel` but with the custom training loop?

Hi @llan-ml,

I tried that and it gives the same wrong behavior. The only way I could make the model converge with a custom training loop was to use the MyKerasModel and train it with it's train_on_batch function.

**Comment by llan-ml** [comment]
There is a bug in your code. In `train_step`, the shape of `y` is `(16,)`, while the shape of `y_pred` is `(16,1)`.

You should try this:
<pre>
def train_step(model, optimizer, x, y):
    y = y[:, tf.newaxis]
    ...
</pre>

**Comment by FelixWiewel** [comment]
@llan-ml Thanks for pointing that out. Indeed this seems to solve the issue. But is this intended behavior? I mean shouldn't tensorflow be able to produce the same results regardless of this difference in shape?

**Comment by llan-ml** [comment]
This is about array/tensor broadcast, similar to numpy. If you want to write customized training loops using low-level apis, you should take care and debug more.

**Comment by FelixWiewel** [comment]
@llan-ml Thanks. I guess the issue can be closed.
